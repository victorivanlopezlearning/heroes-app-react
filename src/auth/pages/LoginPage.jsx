import { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../context/AuthContext';

export const LoginPage = () => {

  const { login } = useContext(AuthContext);

  const navigate = useNavigate();

  const onLogin = () => {

    login('Víctor Iván López');

    navigate('/', {
      replace: true
    });
  };

  return (
    <main className="container d-flex flex-column align-items-center mt-5">
      <h1>Inicia sesión en tu cuenta</h1>

      <button
        className="btn btn-dark mt-2"
        onClick={onLogin}
      >
        Iniciar Sesión
      </button>
    </main>
  )
}