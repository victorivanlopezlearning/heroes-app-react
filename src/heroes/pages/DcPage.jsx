import { HeroList } from '../components';

export const DcPage = () => {
  return (
    <>
      <h1>Dc Comics</h1>

      <HeroList publisher='DC Comics' />
    </>
  )
}