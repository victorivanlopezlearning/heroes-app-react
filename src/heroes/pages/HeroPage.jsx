import { Navigate, useNavigate, useParams } from 'react-router-dom'
import { getHeroById } from '../helpers';

export const HeroPage = () => {

  const { heroId } = useParams();
  const navigate = useNavigate();

  const onNavigateBack = () => {
    navigate(-1);
  };

  const hero = getHeroById(heroId);

  if (!hero) {
    return <Navigate to='/marvel' />
  }

  const { id, superhero, alter_ego, characters, first_appearance, publisher } = hero;

  const heroImageUrl = `/assets/heroes/${id}.jpg`;

  return (
    <div className='row mt-5'>
      <div className="col-md-4">
        <img
          className='img-thumbnail'
          src={heroImageUrl}
          alt={`Imagen ${superhero}`}
        />
      </div>

      <div className="mt-2 p-3 mt-md-0 col-md-8">
        <h2>{superhero}</h2>
        <ul className='list-group list-group-flush'>
          <li className='list-group-item fw-bold'>
            Alter ego:
            <span className='fw-normal'> {alter_ego}</span>
          </li>
          <li className='list-group-item fw-bold'>
            Publisher:
            <span className='fw-normal'> {publisher}</span>
          </li>
          <li className='list-group-item fw-bold'>
            First Appearance:
            <span className='fw-normal'> {first_appearance}</span>
          </li>
        </ul>

        <h5 className='mt-3 fw-bold'>Characters</h5>
        <p>{characters}</p>

        <button
          className='btn btn-dark'
          onClick={onNavigateBack}
        >
          Regresar
        </button>
      </div>
    </div>
  )
}