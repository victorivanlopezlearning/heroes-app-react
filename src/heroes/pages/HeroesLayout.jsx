import { Outlet } from 'react-router-dom';
import { Navbar } from '../../ui';

export const HeroesLayout = () => {
  return (
    <>
      <header>
        <Navbar />
      </header>

      <main className="container mt-5">
        <Outlet />
      </main>
    </>
  )
}
