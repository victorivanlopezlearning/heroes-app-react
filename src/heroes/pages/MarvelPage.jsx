import { HeroList } from '../components';

export const MarvelPage = () => {
  return (
    <>
      <h1>Marvel Comics</h1>

      <HeroList publisher='Marvel Comics' />
    </>
  )
}