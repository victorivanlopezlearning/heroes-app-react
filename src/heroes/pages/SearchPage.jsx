import { useMemo } from 'react';
import { useSearchParams } from 'react-router-dom';
import { useForm } from '../../hooks/useForm'
import { getHeroesByName } from '../helpers';
import { HeroCard } from '../components/HeroCard';

export const SearchPage = () => {

  const [searchParams, setSearchParams] = useSearchParams();
  const heroSearch = useMemo(() => searchParams.get('q') || '', [searchParams]);

  const { search, onInputChange } = useForm({
    search: heroSearch
  });

  const heroes = useMemo(() => getHeroesByName(heroSearch), [heroSearch]);
  const showSearch = useMemo(() => heroSearch.length > 0, [heroSearch]);
  const showError = useMemo(() => (heroSearch.length > 0 && heroes.length === 0), [heroSearch, heroes]);

  const onSearchSubmit = (e) => {
    e.preventDefault();
    if ([search.trim()].includes('')) return;

    setSearchParams(`?q=${search}`);
  }

  return (
    <>
      <h1>BÃºsqueda</h1>

      <div className="row my-4">
        <div className="col-12 col-md-4 mb-4 mb-md-0">
          <form
            onSubmit={onSearchSubmit}
          >
            <div className="mb-3">
              <input
                type="text"
                id="search"
                name="search"
                placeholder="Nombre del heroe"
                autoComplete="off"
                className="form-control"
                value={search}
                onChange={onInputChange}
              />
            </div>

            <button type="submit" className="btn btn-dark">Buscar</button>
          </form>
        </div>

        <div className="col-12 col-md-8">

          {
            showSearch 
              ? heroes.map(hero => <HeroCard key={hero.id} {...hero} />)
              : (
                <div className="alert alert-secondary" role="alert">
                Busca un heroe
              </div>
              )
            
          }

          {
            showError && (
              <div className="alert alert-danger" role="alert">
                <span className="fw-bold">{heroSearch}</span> no encontrado. Favor de buscar otro heroe
              </div>
            )
          }
        </div>
      </div>
    </>
  )
}
