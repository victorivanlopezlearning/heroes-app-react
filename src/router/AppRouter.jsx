import { Navigate } from 'react-router-dom';
import { HeroesRoutes } from '../heroes/routes/HeroesRoutes';
import { LoginPage } from '../auth';
import { HeroesLayout } from '../heroes/pages/HeroesLayout';

export const AppRouter = [
  {
    path: '/',
    element: <HeroesLayout />,
    children: HeroesRoutes,
  },
  {
    path: 'login',
    element: <LoginPage />
  }
]