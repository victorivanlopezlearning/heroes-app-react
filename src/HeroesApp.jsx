import { RouterProvider, createBrowserRouter } from 'react-router-dom';
import { AppRouter } from './router/AppRouter';

const router = createBrowserRouter(AppRouter);

export const HeroesApp = () => {
  return <RouterProvider router={router} />
}