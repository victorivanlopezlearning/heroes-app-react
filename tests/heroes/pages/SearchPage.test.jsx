import { fireEvent, render, screen } from '@testing-library/react';
import { SearchPage } from '../../../src/heroes/pages/SearchPage';
import { RouterProvider, createMemoryRouter, useSearchParams } from 'react-router-dom';
import { AppRouter } from '../../../src/router/AppRouter';
import { AuthContext } from '../../../src/auth';

const mockSetSearchParams = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useSearchParams: () => [jest.requireActual("react-router-dom").useSearchParams()[0], mockSetSearchParams]
}));

describe('Tests in <SearchPage />', () => {

  beforeEach(() => jest.clearAllMocks());

  test('should show with default values', () => {

    const router = createMemoryRouter(AppRouter, {
      initialEntries: ['/search'],
    });

    const { container } = render(
      <AuthContext.Provider value={{ logged: true }}>
        <RouterProvider router={router}>
          <SearchPage />
        </RouterProvider>
      </AuthContext.Provider>
    );

    expect(container).toMatchSnapshot();
  });

  test('should show a hero and the input value of the queryString', () => {

    const heroSearched = 'batman';

    const router = createMemoryRouter(AppRouter, {
      initialEntries: [`/search?q=${heroSearched}`],
    });

    render(
      <AuthContext.Provider value={{ logged: true }}>
        <RouterProvider router={router}>
          <SearchPage />
        </RouterProvider>
      </AuthContext.Provider>
    );

    const input = screen.getByRole('textbox');
    expect(input.value).toBe(heroSearched);
  });

  test('should show an error if a hero is not found', () => {

    const heroSearched = 'batman1234';

    const router = createMemoryRouter(AppRouter, {
      initialEntries: [`/search?q=${heroSearched}`],
    });

    render(
      <AuthContext.Provider value={{ logged: true }}>
        <RouterProvider router={router}>
          <SearchPage />
        </RouterProvider>
      </AuthContext.Provider>
    );

    const alert = screen.getByRole('alert');

    expect(alert.textContent).toBe(`${heroSearched} no encontrado. Favor de buscar otro heroe`);
  });

  test('should call the setSearchParams function', () => {

    const heroSearched = 'batman';

    const router = createMemoryRouter(AppRouter, {
      initialEntries: [`/search`],
    });

    render(
      <AuthContext.Provider value={{ logged: true }}>
        <RouterProvider router={router}>
          <SearchPage />
        </RouterProvider>
      </AuthContext.Provider>
    );

    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: heroSearched } });

    const form = screen.getByRole('form');
    fireEvent.submit(form);

    expect(mockSetSearchParams).toHaveBeenCalled();
    expect(mockSetSearchParams).toHaveBeenCalledWith(`?q=${heroSearched}`);
  });
});