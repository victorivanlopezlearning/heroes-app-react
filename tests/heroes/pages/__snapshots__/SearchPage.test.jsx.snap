// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <SearchPage /> should show with default values 1`] = `
<div>
  <header>
    <nav
      class="navbar navbar-expand-lg navbar-dark bg-dark py-3"
    >
      <div
        class="container-fluid container"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          Asociaciones
        </a>
        <button
          aria-controls="navbarToggle"
          aria-expanded="false"
          aria-label="Toggle navigation"
          class="navbar-toggler"
          data-bs-target="#navbarToggle"
          data-bs-toggle="collapse"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="collapse navbar-collapse"
          id="navbarToggle"
        >
          <ul
            class="navbar-nav me-auto mb-2 mb-lg-0"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link "
                href="/marvel"
              >
                Marvel
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link "
                href="/dc"
              >
                DC
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                aria-current="page"
                class="nav-link active"
                href="/search"
              >
                Buscar
              </a>
            </li>
          </ul>
          <div
            class="d-flex align-items-center justify-content-between justify-content-md-start"
          >
            <span
              class="text-primary me-md-3"
            />
            <button
              class="btn btn-outline-light"
            >
              Cerrar Sesión
            </button>
          </div>
        </div>
      </div>
    </nav>
  </header>
  <main
    class="container mt-5"
  >
    <h1>
      Búsqueda
    </h1>
    <div
      class="row my-4"
    >
      <div
        class="col-12 col-md-4 mb-4 mb-md-0"
      >
        <form
          role="form"
        >
          <div
            class="mb-3"
          >
            <input
              autocomplete="off"
              class="form-control"
              id="search"
              name="search"
              placeholder="Nombre del heroe"
              type="text"
              value=""
            />
          </div>
          <button
            class="btn btn-dark"
            type="submit"
          >
            Buscar
          </button>
        </form>
      </div>
      <div
        class="col-12 col-md-8"
      >
        <div
          class="row g-3"
        >
          <div
            class="alert alert-secondary"
            role="alert"
          >
            Busca un heroe
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
